tools/agarib/agarib.1
.\"
.\" Copyright (c) 2022 Julien Nadeau Carriere <vedge@csoft.net>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd OCTOBER 15, 2022
.Dt AGARIB 1
.Os
.ds vT Agar API Reference
.ds oS Agar 1.7
.Sh NAME
.Nm agarib
.Nd Agar Interface Builder
.Sh SYNOPSIS
.Nm agarib
.Op Fl hvD
.Op Fl d Ar agar-driver
.Op Ar file
.Sh DESCRIPTION
.Nm
is the Agar Interface Builder.
The command-line options are as follows:
.Bl -tag -width Ds
.It Fl h
Print usage and exit.
.It Fl v
Print version number and exit.
.It Fl d
Specify an alternate Agar driver / backend (see
.Dq AVAILABLE DRIVERS
in
.Xr AG_InitGraphics 3
for details; some drivers accept optional arguments).
.It Fl D
Enable debugging mode.
.El
.Sh ENVIRONMENT
.Bl -tag -width "LANG "
.It Dv LANG
ISO language code determining which translation should be used.
.El
.\" .Sh FILES
.Sh SEE ALSO
.Xr AG_Intro 3


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.7.1] - Unreleased
### Added
- Added [cmake](https://cmake.org) support.
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type. Fix widget geometries not updating when switching to a different Type filter.

## [1.7.0] - 2023-05-02
### Added
- [**sdl2fb**](https://libagar.org/man3/AG_DriverSDL2FB): New driver for SDL 2.0 (single-window; frame-buffer mode).
- [**sdl2gl**](https://libagar.org/man3/AG_DriverSDL2GL): New driver for SDL 2.0 (single-window; OpenGL mode).
- [**sdl2mw**](https://libagar.org/man3/AG_DriverSDL2MW): New driver for SDL 2.0 (multi-window; OpenGL mode).
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Added support for X Input Methods. Thanks sr20det!
- [**AG_Button**](https://libagar.org/man3/AG_Button): New option `AG_BUTTON_CROP` (crop label surface to fit contents). Useful for condensing or removing typographical spacings so that individual glyphs (for example "Geometrical Shapes") can be made to align precisely.
- [**AG_Color**](https://libagar.org/man3/AG_Color): Added compression / decompression macros for 4/12/24/48-bit values.
- [**AG_Driver**](https://libagar.org/man3/AG_Driver): New operation: `setMouseAutoCapture` (enable, disable or reset mouse auto-capture on button-down).
- [**AG_DriverSw**](https://libagar.org/man3/AG_DriverSw): New option `clampOnResize` (clamp active Agar windows against new display size).
- [**AG_Error**](https://libagar.org/man3/AG_Error): New function `AG_Debug2()` (output debug message when debug level is >= 2).
- [**AG_FontBf**](https://libagar.org/man3/AG_FontBf): New bitmap font engine for the `.agbf` font format.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Implemented keyboard shortcuts. Added `AG_MENU_NO_BOOL_MSG` and `AG_MENU_FAST_BOOL_MSG` options.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Improved the performance of the `AG_MenuView` renderer by caching state-dependent opaque labels (embedding any icons), so blending is no longer required by `draw()`. Improved appearance of `AG_MenuView`.
- [**AG_Object**](https://libagar.org/man3/AG_Object): Introduce numerical class IDs and `*_ISA()` macros.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_ObjectFreeChildrenOfType()`.
- [**AG_Radio**](https://libagar.org/man3/AG_Radio): New `HOMOGENOUS` option to divide space equally between items. Thanks Stephen!
- [**AG_StyleSheet**](https://libagar.org/man3/AG_StyleSheet): New CSS selectors `E > F`. Child widgets are now addressable by instance name (e.g., `AG_Combo > "trigger"`) or class name (e.g., `AG_Combo > AG_Button`).
- [**AG_StyleSheet**](https://libagar.org/man3/AG_StyleSheet): New CSS selector conditionals `width`, `height` and `zoom`. Widgets can be styled based on their current geometry or zoom level.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): New field `Lpadding` (scaline start padding). Allows fast cropping of lines from the left.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Make it possible to increment the `pixels` pointer to fast crop lines from the top. Added `pixelsBase` pointer.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Added standard palettes for 1/2/4/8-bit modes. New Indexed surfaces are now initialized with a standard palette by default.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Added support for 40- and 48-bpp surfaces.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Introduce optimized blitter tables.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): New function `AG_PixelFormatMaximum()` to return the maximum possible pixel value for a given format.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): New function `AG_SurfaceBlend8()` for blending 1/2/4/8-bpp surfaces against an `AG_Color` using the closest approximation available.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Embed 4 general-purpose 16-bit guides. Guides are useful for typography, graphics and other applications.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Make the `AG_SURFACE_TRACE` flag useful for tracing surface operations such as blits, copies / conversions and mapping to widgets (needs `--enable-debug-surfaces`).
- [**AG_Text**](https://libagar.org/man3/AG_Text): Expand the list of standard font sizes to 32 sizes. Added fractional sizes between 7.0 and 13.0.
- [**AG_Text**](https://libagar.org/man3/AG_Text): Expanded font adjustments to include subpixel alignment corrections for all 32 standard sizes. The table now covers over 150 popular open-source fonts. Adjustments allow different fonts to align with pixel-perfect accuracy.
- [**AG_Text**](https://libagar.org/man3/AG_Text): Added OS/2 metrics and Unicode range information to the `AG_Font` structure.
- [**AG_Text**](https://libagar.org/man3/AG_Text): Utilize OS/2 metrics `typoAscender` and `typoDescender` where available.
- [**AG_Text**](https://libagar.org/man3/AG_Text): New macros `AGSI_IS_*(c)` for testing whether a character lies within a given Unicode range.

tools/agarib/agarib.c
/*
 * Copyright (c) 2022-2023 Julien Nadeau Carriere <vedge@csoft.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "agarib.h"

#include <config/datadir.h>
#include <config/version.h>

/*
 * The Agar Interface Builder.
 *
 * Allows the developer to graphically construct sets of Agar GUI elements
 * and save them to a portable binary (*.ag) format which can be later read
 * by the developer's application.
 */

AG_Window *ibMain = NULL;				/* Main window */
AG_Menu *ibMainMenu = NULL;				/* Main menu */
AG_User *ibUser = NULL;					/* Effective user */
AG_Console *ibConsole = NULL;				/* Text console */
AG_Textbox *ibInput = NULL;				/* Input command */

static void
New_GUI(AG_Event *event)
{
	AG_Window *win;

	if ((win = AG_WindowNew(0)) == NULL) {
		AG_Verbose("%s\n", AG_GetError());
		return;
	}
	AG_WindowSetCaption(win, _("Untitled - Agar Interface Builder"));

	
	AG_WindowSetGeometryAligned(win, AG_WINDOW_MC, 320, 240);
	AG_WindowShow(win);
}

static void
Load_GUI(AG_Event *event)
{
}

static void
Save_GUI(AG_Event *event)
{
}

static void
Save_GUI_As(AG_Event *event)
{
}

static void
Open_GUI(AG_Event *event)
{
	char path[AG_PATHNAME_MAX];
	AG_Window *win;
	AG_FileDlg *fd;
	AG_User *u;

	if ((win = AG_WindowNew(0)) == NULL ||
	    (u = AG_GetEffectiveUser()) == NULL) {
		AG_Verbose("%s\n", AG_GetError());
		return;
	}
	AG_WindowSetCaption(win, _("Open GUI File"));
	AG_WindowSetPosition(win, AG_WINDOW_BC, 0);
	AG_WindowSetMinSize(win, 320, 240);

	fd = AG_FileDlgNew(win, AG_FILEDLG_EXPAND | AG_FILEDLG_MASK_EXT);

	Strlcpy(path, u->home, sizeof(path));
	Strlcat(path, "/GUIs", sizeof(path));
	if (AG_FileExists(path)) {
		AG_FileDlgSetDirectoryMRU(fd, "ag-guis-dir", path);
	} else {
		AG_FileDlgSetDirectoryMRU(fd, "ag-guis-dir", u->home);
	}
	AG_FileDlgAddType(fd, _("Agar GUI File"), "*.ag",
	    Load_GUI, "%p", win);

	AG_WindowShow(win);
}

#if defined(AG_DEBUG) && defined(AG_TIMERS)
static void
Run_GUI_Debugger(AG_Event *event)
{
	AG_Window *win;

	if ((win = (AG_Window *) AG_GuiDebugger(agWindowFocused)) == NULL) {
		AG_Verbose("GUI Debugger: %s\n", AG_GetError());
		return;
	}
	AG_WindowShow(win);
}
#endif /* AG_DEBUG and AG_TIMERS */

static void
PrintVersion(void)
{
	AG_AgarVersion av;
	AG_ConsoleLine *cl;
	AG_Color c;
		
	cl = AG_ConsoleMsg(ibConsole, _("Agar Interface Builder, version %s"),
	    VERSION);
	AG_ColorRGB_8(&c, 0,170,170);
	AG_ConsoleMsgColor(cl, &c);

	AG_GetVersion(&av);
	cl = AG_ConsoleMsg(ibConsole, "LibAgar version %d.%d.%d "
	    "(" AGSI_LEAGUE_SPARTAN "%s" AGSI_RST ")",
	    av.major, av.minor, av.patch,
	    av.release);
	AG_ColorRGB_8(&c, 0,170,0);
	AG_ConsoleMsgColor(cl, &c);
}

int
main(int argc, char *argv[])
{
	char *optArg = NULL, *driverSpec = NULL;
	AG_MenuItem *mi;
	AG_Notebook *nb;
	AG_NotebookTab *nt;
	Uint initFlags = AG_VERBOSE;
	int optInd, i, c;

#ifdef ENABLE_NLS
	bindtextdomain("agarib", LOCALEDIR);
	bind_textdomain_codeset("agarib", "UTF-8");
	textdomain("agarib");
#endif
	while ((c = AG_Getopt(argc, argv, "Dqd:?h", &optArg, &optInd)) != -1) {
		switch (c) {
		case 'D':
			agDebugLvl = 1;
			break;
		case 'q':
			agDebugLvl = 0;
			initFlags &= ~(AG_VERBOSE);
			break;
		case 'd':
			driverSpec = optArg;
			break;
		case '?':
		case 'h':
		default:
			printf("Usage: agarib [-Dq] [-d driver] [file ...]\n");
			return (1);
		}
	}
	if (AG_InitCore("agarib", initFlags) == -1) {
		return (1);
	}
	if ((ibUser = AG_GetEffectiveUser()) == NULL)
		goto fail_core;

	if (AG_InitGraphics(driverSpec) == -1)
		goto fail_core;

	if ((ibMain = AG_WindowNew(AG_WINDOW_MAIN)) == NULL) {
		goto fail_gfx;
	}

	AG_WindowSetCaption(ibMain, _("Agar Interface Builder"));

	ibMainMenu = (agDriverSw) ? AG_MenuNewGlobal(0) :
                                    AG_MenuNew(ibMain, AG_MENU_HFILL);

	mi = AG_MenuNode(ibMainMenu->root, ("File"), NULL);
	{
		AG_MenuItem *miTmpl;

		AG_MenuActionKb(mi, _("New GUI"),
		    agIconDoc.s,
		    AG_KEY_N, AG_KEYMOD_CTRL,
		    New_GUI, NULL);

		miTmpl = AG_MenuNode(mi, _("New GUI From Template"), agIconDoc.s);
		{
			char path[AG_FILENAME_MAX];

			AG_MenuActionKb(mi, _("Open..."),
			    agIconLoad.s,
			    AG_KEY_O, AG_KEYMOD_CTRL,
			    Open_GUI, NULL);

			Strlcpy(path, ibUser->home, sizeof(path));
			Strlcat(path, AG_PATHSEP, sizeof(path));
			Strlcat(path, "Templates", sizeof(path));
			Strlcat(path, AG_PATHSEP, sizeof(path));

			if (AG_FileExists(path)) {
				/* TODO */
			} else {
				AG_MenuNode(miTmpl,
				    AG_Printf(_("( Missing template directory "
				              AGSI_LEAGUE_GOTHIC "%s" AGSI_RST " )"),
				              path),
				    NULL);
			}
		}

		AG_MenuActionKb(mi, _("Save"),
		    agIconSave.s,
		    AG_KEY_S, AG_KEYMOD_CTRL,
		    Save_GUI, NULL);

		AG_MenuActionKb(mi, _("Save As..."),
		    agIconSave.s,
		    AG_KEY_S, AG_KEYMOD_SHIFT|AG_KEYMOD_CTRL,
		    Save_GUI_As, NULL);

		AG_MenuSeparator(mi);

		AG_MenuActionKb(mi, _("GUI Debugger..."), agIconMagnifier.s,
		    AG_KEY_F7, 0,
		    Run_GUI_Debugger, NULL);

		AG_MenuSeparator(mi);

		AG_MenuActionKb(mi, _("Quit"), agIconClose.s,
		    AG_KEY_Q, AG_KEYMOD_CTRL,
		    AGWINCLOSE(ibMain));		/* Close main window */
	}

	mi = AG_MenuNode(ibMainMenu->root, ("Help"), NULL);
	{
		AG_MenuActionKb(mi, _("Contents"), agIconDoc.s,
		    AG_KEY_F1, AG_KEYMOD_NONE,
		    NULL, NULL);
		AG_MenuAction(mi, _("About Agar GUI"), agIconSmallArrowRight.s,
		    AG_About, NULL);
	}

	nb = AG_NotebookNew(ibMain, AG_NOTEBOOK_EXPAND);
	nt = AG_NotebookAdd(nb, _("Console"), AG_BOX_VERT);
	{
		AG_Box *hb;

		ibConsole = AG_ConsoleNew(nt, AG_CONSOLE_EXPAND);

		hb = AG_BoxNewHoriz(nt, AG_BOX_HFILL);
		AG_SetFontWeight(hb, "bold");
		AG_SetFontSize(hb, "120%");
		{
			ibInput = AG_TextboxNew(hb,
			    AG_TEXTBOX_EXCL | AG_TEXTBOX_HFILL,
			    "> ");
//			AG_SetEvent(ibInput, "textbox-return", ExecCmd, NULL);
		}

//		AG_ActionFn(ibConsole, "Help", ConsoleHelp, NULL);
//		AG_ActionOnKey(ibConsole, AG_KEY_F1, AG_KEYMOD_ANY, "Help");
//		AG_ActionOnKey(ibConsole, AG_KEY_HELP, AG_KEYMOD_ANY, "Help");

		AG_WidgetFocus(ibInput);
	}
	
	PrintVersion();


	AG_WindowShow(ibMain);
	AG_WindowSetMinSize(ibMain, 320, 200);
	AG_WindowSetGeometryAligned(ibMain, AG_WINDOW_BC, 500, 300);
	AG_EventLoop();

	AG_UserFree(ibUser);
	AG_DestroyGraphics();
	AG_Destroy();
	return (0);
fail_gfx:
	AG_DestroyGraphics();
fail_core:
	AG_Destroy();
	return (1);
}
